File: aiqcomponents/src/main/java/com/aiqfome/aiqcomponents/selector/Selector.java
package com.aiqfome.aiqcomponents.selector;

import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.drawable.Drawable;
import android.util.AttributeSet;
import android.util.Log;
import android.util.TypedValue;
import android.view.LayoutInflater;
import android.widget.LinearLayout;

import androidx.annotation.Nullable;
import androidx.core.content.ContextCompat;
import androidx.vectordrawable.graphics.drawable.VectorDrawableCompat;

import com.aiqfome.aiqcomponents.R;
import com.google.android.material.textfield.TextInputEditText;
import com.google.android.material.textfield.TextInputLayout;

public class Selector extends LinearLayout {

    private static final String TAG = Selector.class.getSimpleName();

    private SelectorController controller;

    private LinearLayout root;
    private TextInputLayout inputLayout;
    private TextInputEditText editText;

    public Selector(Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);

        LayoutInflater inflater = LayoutInflater.from(context);
        inflater.inflate(R.layout.selector, this, true);

        root = findViewById(R.id.root);
        inputLayout = findViewById(R.id.input_layout);
        editText = findViewById(R.id.et_input);
        editText.setKeyListener(null);
        Drawable icChevronDown = VectorDrawableCompat.create(context.getResources(),
                R.drawable.ic_chevron_down, context.getTheme());
        editText.setCompoundDrawablesWithIntrinsicBounds(null, null, icChevronDown, null);

        setupAttrs(context, attrs);

        setOnClickListener(onClickListener());
        root.setOnClickListener(onClickListener());
        inputLayout.setOnClickListener(onClickListener());
        editText.setOnClickListener(onClickListener());

        TypedValue outValue = new TypedValue();
        getContext().getTheme().resolveAttribute
                (android.R.attr.selectableItemBackground, outValue, true);

        getRootView().setBackgroundResource(outValue.resourceId);
    }

    private void setupAttrs(Context context, @Nullable AttributeSet attrs) {
        if (attrs == null) return;

        TypedArray styledAttributes = context.obtainStyledAttributes(attrs, R.styleable.Selector);

        int attrsSize = styledAttributes.getIndexCount();

        for (int i = 0; i < attrsSize; i++) {
            int attr = styledAttributes.getIndex(i);

            if (attr == R.styleable.Selector_selectorBackgroundColor) {
                if (styledAttributes.hasValue(attr)) {
                    int backgroundColor = styledAttributes
                            .getColor(attr, getResources().getColor(R.color.colorBackground));

                    editText.setBackgroundColor(backgroundColor);
                }

            } else if (attr == R.styleable.Selector_selectorTextAppearance) {
                int textAppearance = styledAttributes.getResourceId(attr, -1);
                editText.setTextAppearance(context, textAppearance);

            } else if (attr == R.styleable.Selector_selectorTitle) {
                if (styledAttributes.hasValue(attr)) {
                    String hint = styledAttributes.getString(attr);
                    if (hint != null) inputLayout.setHint(hint);
                }

            } else {
                Log.d(TAG, "Unknown attribute for " + getClass().toString() + ": " + attr);
            }
        }

        styledAttributes.recycle();
    }

    private OnClickListener onClickListener() {
        return v -> {
            if (controller != null) controller.show();
            else Log.e(TAG, "no controller found, please setup Selector Component!");
        };
    }

    public void setup(SelectorController controller) {
        controller.setSelector(this);
        this.controller = controller;
    }

    @Override
    public void setEnabled(boolean enabled) {
        super.setEnabled(enabled);
        super.setClickable(enabled);

        inputLayout.setEnabled(enabled);
        inputLayout.setClickable(enabled);

        editText.setEnabled(enabled);
        editText.setClickable(enabled);

        if (!enabled) {
            int color = ContextCompat.getColor(getContext(), R.color.colorDivider);
            editText.setBackgroundColor(color);
        }
    }

    public void setSelectedItem(String itemText) {
        this.editText.setText(itemText);
    }


    public SelectorController getController() {
        return controller;
    }

    public TextInputLayout getInputLayout() {
        return inputLayout;
    }

    public TextInputEditText getEditText() {
        return editText;
    }
}

File: aiqcomponents/src/main/java/com/aiqfome/aiqcomponents/selector/Selector.kt
package com.aiqfome.aiqcomponents.selector

import android.content.Context
import android.graphics.drawable.Drawable
import android.os.Parcel
import android.os.Parcelable
import android.util.AttributeSet
import android.util.Log
import android.util.SparseArray
import android.view.LayoutInflater
import android.view.View.OnClickListener
import android.widget.LinearLayout
import androidx.core.content.ContextCompat
import androidx.core.widget.TextViewCompat
import androidx.vectordrawable.graphics.drawable.VectorDrawableCompat
import com.aiqfome.aiqcomponents.R
import com.aiqfome.aiqcomponents.databinding.SelectorBinding
import com.aiqfome.aiqcomponents.utils.extensions.restoreChildViewStates
import com.aiqfome.aiqcomponents.utils.extensions.saveChildViewStates

class Selector @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = 0,
    defStyleRes: Int = 0,
) : LinearLayout(context, attrs, defStyleAttr, defStyleRes) {
    private var _binding: SelectorBinding = SelectorBinding.inflate(
        LayoutInflater.from(context),
        this,
        true
    )

    val binding: SelectorBinding get() = _binding

    private var controller: SelectorController<*>? = null

    init {
        attrs?.let { attrsSet ->
            val typedArray = context.obtainStyledAttributes(
                attrsSet,
                R.styleable.Selector,
                defStyleAttr,
                defStyleRes
            )

            typedArray.getColor(
                R.styleable.Selector_selectorBackgroundColor,
                ContextCompat.getColor(context, R.color.colorBackground),
            ).run {
                binding.inputLayout.setBackgroundColor(this)
            }

            typedArray.getResourceId(
                R.styleable.Selector_selectorTextAppearance,
                -1
            ).let {
                if (it > 0) {
                    TextViewCompat.setTextAppearance(binding.etInput, it)
                }
            }

            typedArray.getString(R.styleable.Selector_selectorTitle)?.let {
                binding.inputLayout.hint = it
            }

            typedArray.recycle()
        }

        binding.etInput.keyListener = null

        val icChevronDown: Drawable? = VectorDrawableCompat.create(
            context.resources,
            R.drawable.ic_chevron_down,
            context.theme
        )
        binding.etInput.setCompoundDrawablesWithIntrinsicBounds(
            null,
            null,
            icChevronDown,
            null
        )

        setOnClickListener(onClickListener())

        binding.root.setOnClickListener(onClickListener())

        binding.inputLayout.setOnClickListener(onClickListener())
        binding.etInput.setOnClickListener(onClickListener())
    }

    private fun onClickListener(): OnClickListener {
        return OnClickListener {
            controller?.show() ?: run {
                Log.e(TAG, "no controller found, please setup Selector Component!")
            }
        }
    }

    fun setup(controller: SelectorController<*>) {
        controller.setSelector(this)
        this.controller = controller
    }

    override fun setEnabled(enabled: Boolean) {
        super.setEnabled(enabled)
        super.setClickable(enabled)

        binding.inputLayout.isEnabled = enabled
        binding.inputLayout.isClickable = enabled
        binding.etInput.isEnabled = enabled
        binding.etInput.isClickable = enabled

        if (!enabled) {
            val color = ContextCompat.getDrawable(context, R.drawable.rounded_backgroun_disable)
            binding.etInput.background = color
            binding.inputLayout.alpha = 0.6f
        } else binding.inputLayout.alpha = 1f
    }

    fun setSelectedItem(itemText: String?) {
        binding.etInput.setText(itemText)
    }

    companion object {
        private val TAG = Selector::class.java.simpleName
    }

    override fun onSaveInstanceState(): Parcelable {
        return SavedState(super.onSaveInstanceState()).apply {
            childrenStates = saveChildViewStates()
        }
    }

    public override fun onRestoreInstanceState(state: Parcelable) {
        Log.i("SavedState", "onRestoreInstanceState")
        when (state) {
            is SavedState -> {
                super.onRestoreInstanceState(state.superState)
                state.childrenStates?.let { restoreChildViewStates(it) }
            }
            else -> super.onRestoreInstanceState(state)
        }
    }

    override fun dispatchSaveInstanceState(container: SparseArray<Parcelable>) {
        dispatchFreezeSelfOnly(container)
    }

    override fun dispatchRestoreInstanceState(container: SparseArray<Parcelable>) {
        dispatchThawSelfOnly(container)
    }

    internal class SavedState : BaseSavedState {
        internal var childrenStates: SparseArray<Parcelable>? = null

        constructor(superState: Parcelable?) : super(superState)

        @Suppress("UNCHECKED_CAST")
        constructor(source: Parcel) : super(source) {
            Log.i("SavedState", "Reading children children state from sparse array")
            childrenStates = source.readSparseArray<Parcel>(javaClass.classLoader) as SparseArray<Parcelable>?
        }

        @Suppress("UNCHECKED_CAST")
        override fun writeToParcel(out: Parcel, flags: Int) {
            Log.i("SavedState", "Writing children state to sparse array")
            super.writeToParcel(out, flags)
            out.writeSparseArray(childrenStates as SparseArray<Any>)
        }

        companion object {
            @Suppress("UNUSED")
            @JvmField
            val CREATOR = object : Parcelable.Creator<SavedState> {
                override fun createFromParcel(source: Parcel) = SavedState(source)
                override fun newArray(size: Int): Array<SavedState?> = arrayOfNulls(size)
            }
        }
    }
}